src/Headers/NeighbourManager.h

```diff
@@ -0,0 +1,329 @@
+    // Compute distances and the inverse between the current particle and all neighbours here,
+    // for both gather and inactive scatter neibs.
+    //-----------------------------------------------------------------------------------------
+    for (int j=0; j<neiblist.size(); j++) {
+      int jj = neiblist[j] ;
+
+      // Skip if (i) neighbour particle type does not interact hydrodynamically with particle,
+      // (ii) neighbour is a dead (e.g. accreted) particle. Additionally, if do_pair_once is true,
+      /// (iii) same i.d. as current active particle, (iv) neighbour is on lower timestep level (i.e.
+      // timestep is shorter), or (v) neighbour is on same level as current particle but has larger id. value
+      // (to only calculate each pair once).
+      if (hydromask[neibdata[jj].ptype] == false) continue;
+
+      //if (p.iorig==neibdata[jj].iorig) continue;
+
+      if (do_pair_once) {
+
+        const bool need_interaction =
+            (neibdata[jj].flags.is_mirror()) ||
+            (p.level > neibdata[jj].level) ||
+            (p.level == neibdata[jj].level &&
+                p.iorig < neibdata[jj].iorig) ;
+
+        if (not need_interaction) continue;
+      }
+
+      for (int k=0; k<ndim; k++) draux[k] = neibdata[jj].r[k] - rp[k];
+      if (jj<_NPeriodicGhosts) {
+        GhostFinder.NearestPeriodicVector(draux);
+      }
+      drsqd = DotProduct(draux,draux,ndim)+small_number;
+
+      //          if (drsqd <= small_number) continue ;
+
+      // Compute relative position and distance quantities for pair
+      if (drsqd <= hrangesqdi || drsqd <= neibdata[jj].hrangesqd) {
+        culled_neiblist.push_back(jj);
+        if (jj<_NPeriodicGhosts) for (int k=0; k<ndim; k++) neibdata[jj].r[k] = draux[k]+rp[k];
+      }
+    }
+
+    *neiblist_p=&culled_neiblist[0];
+    *neibdata_p=&neibdata[0];
+
+
+
+    return culled_neiblist.size();
+
+  }
+
+  template<class InParticleType>
+  ListLength GetParticleNeibGravity(const InParticleType& p,const Typemask& hydromask, const Typemask& gravmask, int** neiblist_p,
+      int** directlist_p, int** gravlist_p, ParticleType** neibdata_p, const bool do_grav) {
+
+    FLOAT rp[ndim];
+    FLOAT draux[ndim];
+
+    for (int k=0; k<ndim; k++) rp[k] = p.r[k];
+    const FLOAT hrangesqdi = p.hrangesqd;
+
+    const GhostNeighbourFinder<ndim> GhostFinder(*_domain);
+
+    culled_neiblist.clear();
+    gravlist.clear();
+    // Particles that are already in the directlist stay there; we just add the ones that were demoted
+    directlist.resize(_NCellDirectNeib);
+
+    // Go through the hydro neighbour candidates and check the distance. The ones that are not real neighbours
+    // are demoted to the direct list
+    for (int jj=0; jj < neiblist.size(); jj++) {
+      int ii = neiblist[jj];
+
+      // Compute relative position and distance quantities for pair
+      for (int k=0; k<ndim; k++) draux[k] = neibdata[ii].r[k] - rp[k];
+      GhostFinder.NearestPeriodicVector(draux);
```

Comment:
Yeah, I think it's only necessary for those ghastly edge cases - like the large cells.